core/trino-main/src/main/java/io/trino/execution/scheduler/StageTaskSourceFactory.java

```diff
@@ -0,0 +1,705 @@
+        }
+
+        @Override
+        public List<TaskDescriptor> getMoreTasks()
+        {
+            if (finished || closed) {
+                return ImmutableList.of();
+            }
+
+            Map<Integer, Multimap<PlanNodeId, Split>> partitionToSplitsMap = new HashMap<>();
+            Map<Integer, HostAddress> partitionToNodeMap = new HashMap<>();
+            if (!splitSources.isEmpty()) {
+                for (Map.Entry<PlanNodeId, SplitSource> entry : splitSources.entrySet()) {
+                    PlanNodeId scanNodeId = entry.getKey();
+                    SplitSource splitSource = entry.getValue();
+                    BucketNodeMap bucketNodeMap = this.bucketNodeMap
+                            .orElseThrow(() -> new VerifyException("bucket to node map is expected to be present"));
+                    while (!splitSource.isFinished()) {
+                        ListenableFuture<SplitBatch> splitBatchFuture = splitSource.getNextBatch(NOT_PARTITIONED, Lifespan.taskWide(), splitBatchSize);
+
+                        long start = System.nanoTime();
+                        addSuccessCallback(splitBatchFuture, () -> getSplitTimeRecorder.accept(start));
+
+                        SplitBatch splitBatch = getFutureValue(splitBatchFuture);
+
+                        for (Split split : splitBatch.getSplits()) {
+                            int bucket = bucketNodeMap.getBucket(split);
+                            int partition = bucketToPartitionMap.map(map -> map[bucket]).orElse(bucket);
+
+                            if (!bucketNodeMap.isDynamic()) {
+                                partitionToNodeMap.put(partition, bucketNodeMap.getAssignedNode(split).get().getHostAndPort());
+                            }
+
+                            Multimap<PlanNodeId, Split> partitionSplits = partitionToSplitsMap.computeIfAbsent(partition, (p) -> ArrayListMultimap.create());
+                            partitionSplits.put(scanNodeId, split);
+                        }
+
+                        if (splitBatch.isLastBatch()) {
+                            splitSource.close();
+                            break;
+                        }
+                    }
+                }
+            }
+
+            Map<Integer, Multimap<PlanNodeId, ExchangeSourceHandle>> partitionToExchangeSourceHandlesMap = new HashMap<>();
+            for (Map.Entry<PlanNodeId, ExchangeSourceHandle> entry : partitionedExchangeSourceHandles.entries()) {
+                PlanNodeId planNodeId = entry.getKey();
+                ExchangeSourceHandle handle = entry.getValue();
+                int partition = handle.getPartitionId();
+                Multimap<PlanNodeId, ExchangeSourceHandle> partitionSourceHandles = partitionToExchangeSourceHandlesMap.computeIfAbsent(partition, (p) -> ArrayListMultimap.create());
+                partitionSourceHandles.put(planNodeId, handle);
+            }
+
+            // TODO implement small tasks merging
```

Comment:
Do we have an issue for this? Add link to comment

Reply:
We have a work item, but it is unclear what is the priority. Let me remove the comment for now.